src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,316 @@
+    case t: RefinedType => strip(t.underlying)
+    case _ => t
+  }
+
+  def isSubType(t1: Type, t2: Type): Boolean = {
+    val res = t1 <:< t2
+    debug(s"$t1 <: $t2 ? $res")
+    res
+  }
+
+  def isEqualType(t1: Type, t2: Type): Boolean = {
+    val res = t1 =:= t2
+    debug(s"$t1 == $t2 ? $res")
+    res
+  }
+
+  def getSignature(t: Type): List[Type] = {
+    val ktor = t.classSymbol.primaryConstructor.info
+
+    def typeParams(t: Type): List[Type] = t match {
+      case t: RefinedType => t.refinedInfo.asInstanceOf[TypeAlias].alias :: typeParams(t.parent)
+      case _ => Nil
+    }
+
+    val sign = (t match {
+      case t: RefinedType => ktor.appliedTo(typeParams(t))
+      case _ => ktor
+    }).firstParamTypes.map(_.stripTypeVar)
+
+    debug(s"====signature of $t: $sign")
+    sign
+  }
+
+  def getPartitions(t: Type): List[Type] = t match {
+    case OrType(t1, t2) => List(t1, t2)
+    case _ =>
+      val children = t.classSymbol.children.map { c =>
+        if (c is Flags.ModuleClass)
+          c.info.asInstanceOf[ClassInfo].selfType
+        else
+          specializeChild(t, c.info.asInstanceOf[ClassInfo].symbolicTypeRef)
+      }.toList
+
+      debug(s"child of ${t.show}: ${children.map(_.show).mkString(", ")}")
+
+      children
+  }
+
+  // parent is specialized, specialize child as well
+  def specializeChild(p: Type, c: Type): Type = p match {
+    case p: RefinedType => p.wrapIfMember(specializeChild(p.parent, c))
+    case _ => c
+  }
+
+  // abstract sealed types and or types can be decomposed
+  def canDecompose(t: Type): Boolean = {
+    t.typeSymbol.is(Flags.allOf(Flags.Abstract, Flags.Sealed)) || t.isInstanceOf[OrType]
+  }
+
+  def isCaseClass(t: Type): Boolean = t.classSymbol.isClass && t.classSymbol.is(Flags.CaseClass)
+
+  // display spaces
+  def show(s: Space): String = s match {
+    case Empty => ""
+    case Typ(t) =>
+      val sym = t.widen.classSymbol
+
+      if (sym is Flags.ModuleClass)
+        sym.sourceModule.name.show
+      else if (t.isInstanceOf[OrType])
+        "_: " + t.show
+      else if (getSignature(t).isEmpty)
+        "_: " + t.classSymbol.name
+      else if (ctx.definitions.isTupleType(t))
+        "(" + getSignature(t).map(_ => "_").mkString(", ") + ")"
+      else
+        t.classSymbol.name + "(" + getSignature(t).map(_ => "_").mkString(",") + ")"
+    case Kon(t, params) =>
+      if (ctx.definitions.isTupleType(t))
+        "(" + params.map(show _).mkString(", ") + ")"
+      else
+        t.classSymbol.name + "(" + params.map(show _).mkString(", ") + ")"
+    case Or(spaces) => spaces.map(show _).mkString(", ")
+  }
+
+  // does the type t have @unchecked annotation
+  def skipCheck(t: Type): Boolean = t match {
+    case AnnotatedType(t, annot) => ctx.definitions.UncheckedAnnot == annot.symbol
+    case _ => false
+  }
+
+  // widen a type and eliminate anonymous class
+  private def widen(t: Type): Type = t.widen match {
```

Comment:
Looks like this method is adapted from `PatternMatcher#elimAnonymousClass`, instead of copying code it would be better to share it between this and the PatternMatcher, you could for example put `elimAnonymousClass` in `object PatternMatcher`